c_binding/lsm_local_disk.c

```diff
@@ -751,3 +784,182 @@ int lsm_local_disk_vpd83_get(const char *disk_path, char **vpd83,
+{
+    uint16_t supported_list_len = 0;
+    uint16_t i = 0;
+
+    supported_list_len = (vpd_0_data[2] << 8) + vpd_0_data[3];
+
+    for (; (i < supported_list_len) &&
+           ((i + _T10_SPC_VPD_SUP_VPD_PGS_LIST_OFFSET) < vpd_0_len); ++i) {
+        if (page_code == vpd_0_data[i + _T10_SPC_VPD_SUP_VPD_PGS_LIST_OFFSET])
+            return true;
+    }
+    return false;
+}
+
+/*
+ * Assuming err_msg and vpd_data are not NULL, caller should make sure that.
+ */
+static int _sg_io_ioctl_vpd(char *err_msg, int fd, uint8_t page_code,
+                            uint8_t *vpd_data, uint16_t vpd_data_len)
+{
+    int rc = LSM_ERR_OK;
+    uint8_t vpd_00_data[_T10_SPC_VPD_SUP_VPD_PGS_MAX_PAGE_LEN];
+    uint8_t cdb[_T10_SPC_INQUERY_CMD_LEN];
+    struct sg_io_hdr io_hdr;
+    int ioctl_rc = 0;
+    int ioctl_errno = 0;
+    int rc_vpd_00 = 0;
+
+    /* SPC-5 Table 142 - INQUIRY command */
+    cdb[0] = INQUIRY;
+    /* ^ OPERATION CODE */
+    cdb[1] = 1;
+    /* ^  EVPD, VPD INQUIRY require EVPD == 1 */;
+    cdb[2] = page_code & UINT8_MAX;
+    /* ^ PAGE CODE */
+    cdb[3] = (vpd_data_len >> 8 )& UINT8_MAX;
+    /* ^ ALLOCATION LENGTH, MSB */
+    cdb[4] = vpd_data_len & UINT8_MAX;
+    /* ^ ALLOCATION LENGTH, LSB */
+    cdb[5] = 0;
+    /* ^ CONTROL, no need to handle auto contingent allegiance(ACA) */
+
+    memset(&io_hdr, 0, sizeof(struct sg_io_hdr));
+    memset(vpd_data, 0, (size_t) vpd_data_len);
+    io_hdr.interface_id = 'S'; /* 'S' for SCSI generic */
+    io_hdr.cmd_len = _T10_SPC_INQUERY_CMD_LEN;
+    io_hdr.sbp = NULL; /* No need to parse sense data */
+    io_hdr.mx_sb_len = 0;
+    io_hdr.dxfer_direction = SG_DXFER_FROM_DEV;
+    io_hdr.dxfer_len = vpd_data_len;
+    io_hdr.dxferp = vpd_data;
+    io_hdr.cmdp = cdb;
+    io_hdr.timeout = _VPD_QUERY_TMO;
+
+    ioctl_rc = ioctl(fd, SG_IO, &io_hdr);
+    if (ioctl_rc != 0) {
+        ioctl_errno = errno;
+        if (page_code == _T10_SPC_VPD_SUP_VPD_PGS) {
+            _lsm_err_msg_set(err_msg, "Not a SCSI compatible device");
+            return LSM_ERR_NO_SUPPORT;
+        }
+        /* Check whether provided page is supported */
+        rc_vpd_00 = _sg_io_ioctl_vpd(err_msg, fd, _T10_SPC_VPD_SUP_VPD_PGS,
+                                     vpd_00_data,
+                                     _T10_SPC_VPD_SUP_VPD_PGS_MAX_PAGE_LEN);
+        if (rc_vpd_00 != 0) {
+            rc = LSM_ERR_NO_SUPPORT;
+            goto out;
+        }
+
+        if (_is_vpd_page_supported(vpd_00_data,
+                                   _T10_SPC_VPD_SUP_VPD_PGS_MAX_PAGE_LEN,
+                                   page_code) == true) {
+            /* Current VPD page is supported, then it's a library bug */
+            rc = LSM_ERR_LIB_BUG;
+            _lsm_err_msg_set(err_msg,
+                             "BUG: VPD page 0x%02x is supported, "
+                             "but failed with error %d(%s)", ioctl_rc,
+                             strerror(ioctl_errno));
```

Comment:
I'm thinking we should use the thread safe version of strerror.  The library isn't thread safe, but we should be trying to be thread aware.